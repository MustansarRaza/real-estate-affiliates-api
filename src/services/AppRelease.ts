import { AppRelease } from "affiliates-api/models";
import { Op } from "sequelize";

export interface AppVariablesDTO {
  min_version: string;
}
export interface SettingsDTO {
  version: string;
  appVariables: AppVariablesDTO;
}
class AppReleaseService {
  /**
   * Returns mobile application settings
   */
  async get(appName: string): Promise<SettingsDTO | null> {
    const release = await AppRelease.findOne({
      attributes: ["version", "appVariables"],
      where: {
        appName: {
          [Op.eq]: appName,
        },
      },
    });

    return release ? this.serialize(release) : null;
  }
  /**
   * Serializes a queried {@see AppRelease} model instance into a DTO.
   */
  private serialize(release: AppRelease): SettingsDTO {
    return {
      version: release.version.toString(),
      appVariables: {
        min_version: release.appVariables!.min_version,
      },
    };
  }
}
export default AppReleaseService;
