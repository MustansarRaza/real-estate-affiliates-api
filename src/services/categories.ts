import { ListingStatusKey, PlatformKey } from "@pf/taxonomies";
import {
  Address,
  AgencyConfiguration,
  Category,
  Listing,
  ListingStatus,
  Location,
  Platform,
  popularCategoryKeyToIndex,
  Project,
  ProjectAffiliateVisibilityID,
  ProjectStatus,
} from "affiliates-api/models";
import { col, fn, Op } from "sequelize";

interface CategoryAvailabilityDTO {
  availableInCities: CategoryCityAvailabilityDTO[];
  availableInProjects: number[];
}

interface CategoryCityAvailabilityDTO {
  cityID: number;
  unitsCount: number;
}

interface CategoryDTO {
  id: number;
  name: string;
  nameSingular: string;
  orderIndex: number;
  slug: string;
  parentID: number | null;
  popularityIndex: number;
  availableInCities: CategoryCityAvailabilityDTO[];
  availableInProjects: number[];
}

/**
 * Provides access to categories.
 */
class CategoriesService {
  /**
   * Gets a list of all available categories.
   */
  async list(agencyID: number | null): Promise<CategoryDTO[]> {
    let platformId = null;

    if (agencyID) {
      const agencyConfig = await AgencyConfiguration.findOne({
        attributes: ["platformId"],
        where: { agencyId: { [Op.eq]: agencyID } },
        raw: true,
      });
      platformId = agencyConfig?.platformId;
    }
    if (!platformId) {
      const platform = await Platform.findOne({
        attributes: ["id"],
        where: { platformKey: { [Op.eq]: PlatformKey.ZAMEEN } },
        raw: true,
      });
      platformId = platform.id;
    }

    const categories = await Category.findAll({
      where: { platformId: { [Op.eq]: platformId } },
      order: [
        ["type_order", "ASC"],
        ["type_id", "ASC"],
      ],
    });

    const categoryCitiesAndProjectsMap = await this.getCategoryCitiesAndProjectsMap(agencyID);

    return categories.map((category: Category) => ({
      id: category.id,
      name: category.name,
      nameSingular: category.nameSingular,
      orderIndex: category.orderIndex,
      slug: category.slug,
      parentID: category.parentID,
      popularityIndex: popularCategoryKeyToIndex(category.key),
      availableInCities: categoryCitiesAndProjectsMap[category.id]?.availableInCities || [],
      availableInProjects: categoryCitiesAndProjectsMap[category.id]?.availableInProjects || [],
    }));
  }

  /**
   * Gets a map of cities and Projects against a category in which
   * that category is available.
   */
  private async getCategoryCitiesAndProjectsMap(
    agencyID: number | null
  ): Promise<{
    [propertTypeId: number]: CategoryAvailabilityDTO;
  }> {
    const categoriesAvailablityResult = await Listing.findAll({
      attributes: [[fn("COUNT", "id"), "Count"]],
      include: [
        {
          model: ListingStatus,
          attributes: [],
          where: {
            key: {
              [Op.eq]: ListingStatusKey.AVAILABLE,
            },
          },
          required: true,
        },
        {
          model: Category,
          attributes: ["id"],
          required: true,
        },
        {
          model: Project,
          required: true,
          attributes: ["id"],
          where: {
            status: {
              [Op.eq]: ProjectStatus.ON,
            },
            affiliateVisibility: {
              [Op.eq]: ProjectAffiliateVisibilityID.VISIBLE,
            },
            ...(agencyID && {
              agencyID: {
                [Op.eq]: agencyID,
              },
            }),
          },
          include: [
            {
              model: Address,
              required: true,
              attributes: [],
              include: [
                {
                  model: Location,
                  attributes: ["id"],
                  required: true,
                },
              ],
            },
          ],
        },
      ],
      group: [
        col("Category.type_id"),
        col("Project.Address.Location.title"),
        col("Project.project_id"),
      ],
      raw: true,
    });

    const categoryCitiesAndProjectsMap: { [propertTypeId: number]: CategoryAvailabilityDTO } = {};
    categoriesAvailablityResult.forEach((row: any) => {
      const propertyTypeId = row["Category.id"];
      const projectID = row["Project.id"];
      const cityID = row["Project.Address.Location.id"];
      const unitsCount = row["Count"];

      if (categoryCitiesAndProjectsMap.hasOwnProperty(propertyTypeId)) {
        categoryCitiesAndProjectsMap[propertyTypeId].availableInProjects.push(projectID);
        const index = categoryCitiesAndProjectsMap[propertyTypeId].availableInCities
          .map((city: CategoryCityAvailabilityDTO) => city.cityID)
          .indexOf(cityID);
        if (index === -1) {
          categoryCitiesAndProjectsMap[propertyTypeId].availableInCities.push({
            cityID,
            unitsCount,
          });
        } else {
          categoryCitiesAndProjectsMap[propertyTypeId].availableInCities[
            index
          ].unitsCount += unitsCount;
        }
      } else {
        categoryCitiesAndProjectsMap[propertyTypeId] = {
          availableInProjects: [projectID],
          availableInCities: [{ cityID, unitsCount }],
        };
      }
    });

    return categoryCitiesAndProjectsMap;
  }
}

export default CategoriesService;
