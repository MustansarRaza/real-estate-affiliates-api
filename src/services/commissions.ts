import { AffiliateCommissionType } from "@pf/taxonomies";
import { AffiliateCommission } from "affiliates-api/models";
import { maxBy, minBy } from "lodash";
import { Op, WhereOptions } from "sequelize";

export interface ProjectCommissionDTO {
  max: number | null;
  min: number | null;
}

export interface AffiliateCommissionDTO {
  [commissionType: string]: {
    [propertyTypeId: number]: AffiliateCommission[];
  };
}

export interface AffiliateProjectCommissionDTO {
  [projectID: number]: AffiliateCommissionDTO;
}

type ByProjectIDLookup = {
  projectIDs: number[];
};

type ByAffiliateTypeIDLookup = {
  affiliateTypeId: number | null;
};

type ByFiltersLookup = {
  createdBy?: number;
  propertyTypeId?: number;
  commissionType?: AffiliateCommissionType;
};

type AffiliateCommissionLookupOptions = ByProjectIDLookup &
  ByAffiliateTypeIDLookup &
  ByFiltersLookup;

class CommissionsService {
  async getDetailsFor(
    options: AffiliateCommissionLookupOptions
  ): Promise<{ [projectID: number]: AffiliateCommission[] }> {
    const unformattedDate = new Date();
    unformattedDate.setHours(0, 0, 0, 0);
    const currentDate = unformattedDate.toISOString();

    const commissionDetails = await AffiliateCommission.findAll({
      where: {
        [Op.and]: this.createFiltersForAffiliateCommission(options),
        [Op.or]: [
          {
            startDate: {
              [Op.is]: null,
            },
            endDate: {
              [Op.is]: null,
            },
          },
          {
            startDate: {
              [Op.not]: null,
              [Op.lte]: currentDate,
            },
            endDate: {
              [Op.is]: null,
            },
          },
          {
            startDate: {
              [Op.is]: null,
            },
            endDate: {
              [Op.not]: null,
              [Op.gte]: currentDate,
            },
          },
          {
            startDate: {
              [Op.not]: null,
              [Op.lte]: currentDate,
            },
            endDate: {
              [Op.not]: null,
              [Op.gte]: currentDate,
            },
          },
        ],
      },
      raw: true,
    });

    const projectAndCommissionsMapping: { [projectID: number]: AffiliateCommission[] } = {};
    commissionDetails.forEach((row: AffiliateCommission) => {
      if (projectAndCommissionsMapping.hasOwnProperty(row.projectId))
        projectAndCommissionsMapping[row.projectId].push(row);
      else projectAndCommissionsMapping[row.projectId] = [row];
    });

    return projectAndCommissionsMapping;
  }

  async getCommissionsBreakdownForAffiliate(
    options: AffiliateCommissionLookupOptions
  ): Promise<AffiliateProjectCommissionDTO> {
    const projectAndCommissionsMapping = await this.getDetailsFor(options);

    const avaiableCommissionProjectIDs = Object.keys(projectAndCommissionsMapping);

    const affiliateCommissionBreakdown: AffiliateProjectCommissionDTO = {};
    avaiableCommissionProjectIDs.forEach((projectID) => {
      const projectCommissions = projectAndCommissionsMapping[parseInt(projectID, 10)] || [];

      projectCommissions.forEach((commission) => {
        if (affiliateCommissionBreakdown.hasOwnProperty(commission.projectId))
          if (
            affiliateCommissionBreakdown[commission.projectId].hasOwnProperty(
              commission.commissionType
            )
          )
            if (
              affiliateCommissionBreakdown[commission.projectId][
                commission.commissionType
              ].hasOwnProperty(commission.propertyTypeId)
            )
              affiliateCommissionBreakdown[commission.projectId][commission.commissionType][
                commission.propertyTypeId
              ].push(commission);
            else
              affiliateCommissionBreakdown[commission.projectId][commission.commissionType][
                commission.propertyTypeId
              ] = [commission];
          else
            affiliateCommissionBreakdown[commission.projectId][commission.commissionType] = {
              [commission.propertyTypeId]: [commission],
            };
        else
          affiliateCommissionBreakdown[commission.projectId] = {
            [commission.commissionType]: { [commission.propertyTypeId]: [commission] },
          };
      });
    });

    return affiliateCommissionBreakdown;
  }

  async getProjectCommissionStats(
    options: AffiliateCommissionLookupOptions
  ): Promise<Map<number, ProjectCommissionDTO>> {
    const projectAndCommissionsMapping = await this.getDetailsFor(options);

    const projectCommissionStats = new Map<number, ProjectCommissionDTO>();
    Object.keys(projectAndCommissionsMapping).forEach((projectID) => {
      const maxCommission = maxBy(
        projectAndCommissionsMapping[+projectID],
        (affiliateCommission) =>
          affiliateCommission.commission + affiliateCommission.classifiedCredits!
      );
      const minCommission = minBy(
        projectAndCommissionsMapping[+projectID],
        (affiliateCommission) =>
          affiliateCommission.commission + affiliateCommission.classifiedCredits!
      );

      return projectCommissionStats.set(+projectID, {
        max: maxCommission?.commission! + maxCommission?.classifiedCredits!,
        min: minCommission?.commission! + minCommission?.classifiedCredits!,
      });
    });

    return projectCommissionStats;
  }

  private createFiltersForAffiliateCommission(
    options: AffiliateCommissionLookupOptions
  ): WhereOptions[] {
    const filters: WhereOptions[] = [
      {
        affiliateTypeId: {
          [Op.eq]: options.affiliateTypeId,
        },
        projectId: {
          [Op.in]: options.projectIDs,
        },
      },
    ];

    if (options?.createdBy) {
      filters.push({
        createdBy: {
          [Op.eq]: options.createdBy,
        },
      });
    }
    if (options?.propertyTypeId) {
      filters.push({
        propertyTypeId: {
          [Op.eq]: options.propertyTypeId,
        },
      });
    }
    if (options?.commissionType) {
      filters.push({
        commissionType: {
          [Op.eq]: options.commissionType,
        },
      });
    }

    return filters;
  }
}

export default CommissionsService;
