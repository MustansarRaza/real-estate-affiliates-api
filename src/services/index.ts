import AppRelease from "./AppRelease";
import CategoriesService from "./categories";
import ClientsService, { CreateClientDTO as CreateClientDTOType } from "./clients";
import CommissionsService from "./commissions";
import LeadsService, { CreateLeadDTO as CreateLeadDTOType } from "./leads";
import ListingsService, { ListingDTO as ListingDTOType } from "./listings";
import LocationsService from "./locations";
import MeService, { CreateMeDTO as CreateMeDTOType, MeDTO as MeDTOType } from "./me";
import ProfileService from "./profile";
import ProjectsService from "./projects";
import ListingReservationService, { ListingReservationOptions } from "./reservations";
import TasksService from "./tasks";
import TaskTypesService from "./taskTypes";

export interface ServiceMesh {
  profile: ProfileService;
  me: MeService;
  clients: ClientsService;
  tasks: TasksService;
  taskTypes: TaskTypesService;
  leads: LeadsService;
  projects: ProjectsService;
  listings: ListingsService;
  locations: LocationsService;
  categories: CategoriesService;
  reservations: ListingReservationService;
  commissions: CommissionsService;
  appRelease: AppRelease;
}

const useServices = (): ServiceMesh => {
  const locations = new LocationsService();
  const profile = new ProfileService({ locations });
  const me = new MeService({ profile, locations });
  const clients = new ClientsService({ me, locations });
  const taskTypes = new TaskTypesService();
  const leads = new LeadsService({ me, taskTypes });
  const tasks = new TasksService({ leads, taskTypes });
  const listings = new ListingsService({ me });
  const commissions = new CommissionsService();
  const projects = new ProjectsService({ me, listings, commissions, locations });
  const categories = new CategoriesService();
  const reservations = new ListingReservationService({ me, listings });
  const appRelease = new AppRelease();

  return {
    profile,
    me,
    clients,
    leads,
    tasks,
    taskTypes,
    projects,
    listings,
    locations,
    categories,
    reservations,
    commissions,
    appRelease,
  };
};

export {
  useServices,
  ClientsService,
  LeadsService,
  MeService,
  ProfileService,
  ProjectsService,
  ListingsService,
  TasksService,
  TaskTypesService,
  LocationsService,
  CategoriesService,
  ListingReservationService,
  CommissionsService,
  AppRelease,
};

export type CreateClientDTO = CreateClientDTOType;
export type CreateLeadDTO = CreateLeadDTOType;
export type CreateMeDTO = CreateMeDTOType;
export type CreateReservationDTO = ListingReservationOptions;
export type ListingDTO = ListingDTOType;
export type MeDTO = MeDTOType;
