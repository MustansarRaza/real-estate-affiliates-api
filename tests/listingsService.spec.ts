import {
  Address,
  AreaUnit,
  AreaUnitID,
  ListingAgainstType,
  ListingAllocationID,
  ListingCommissionType,
  ListingStatusID,
  ListingVisibilityID,
  Purpose,
  PurposeID,
} from "affiliates-api/models";
import { ListingDTO, ListingsService } from "affiliates-api/services";
import each from "jest-each";
import { loadSequelizeFixtures, useMockedServices } from "./support";
import {
  createListing,
  createListingImage,
  createMediaBankImage,
  ModelFixtureConstants,
  ModelFixtureNames,
} from "./support/data";

describe("ListingsService", () => {
  const mockedServices = useMockedServices();
  const service = new ListingsService(mockedServices);

  beforeEach(async () => {
    await loadSequelizeFixtures(
      ModelFixtureNames.COUNTRIES,
      ModelFixtureNames.LOCATIONS,
      ModelFixtureNames.ADDRESSES,
      ModelFixtureNames.CATEGORIES,
      ModelFixtureNames.PROJECTS,
      ModelFixtureNames.LISTING_STATUSES
    );
    jest.resetAllMocks();
  });

  describe("list", () => {
    it("lists all available listings if no filters", async () => {
      const listing1 = await createListing({ againstType: ListingAgainstType.PROJECT });
      const listing2 = await createListing({ againstType: ListingAgainstType.LEAD });

      const listings = await service.list({
        agencyID: ModelFixtureConstants.verifiedAffiliateAgencyID,
      });
      expect(listings).toHaveLength(2);
      expect(listings[0].id).toBe(listing2.id);
      expect(listings[1].id).toBe(listing1.id);
    });

    it("only lists listings linked to the specified agency", async () => {
      await createListing({ agencyID: 99, againstType: ListingAgainstType.PROJECT });
      const listing2 = await createListing({
        agencyID: ModelFixtureConstants.verifiedAffiliateAgencyID,
        againstType: ListingAgainstType.LEAD,
      });

      const listings = await service.list({
        agencyID: ModelFixtureConstants.verifiedAffiliateAgencyID,
      });
      expect(listings).toHaveLength(1);
      expect(listings[0].id).toBe(listing2.id);
    });

    it("applies againstType filter if applied", async () => {
      await createListing({ againstType: ListingAgainstType.PROJECT });
      const listing2 = await createListing({ againstType: ListingAgainstType.LEAD });

      const listings = await service.list({
        agencyID: ModelFixtureConstants.verifiedAffiliateAgencyID,
        againstType: ListingAgainstType.LEAD,
      });
      expect(listings).toHaveLength(1);
      expect(listings[0].id).toBe(listing2.id);
    });

    it("applies againstID filter if applied", async () => {
      await createListing({ againstID: 1 });
      const listing2 = await createListing({ againstID: 2 });

      const listings = await service.list({
        agencyID: ModelFixtureConstants.verifiedAffiliateAgencyID,
        againstID: 2,
      });
      expect(listings).toHaveLength(1);
      expect(listings[0].id).toBe(listing2.id);
    });

    it("only lists active listings", async () => {
      const activeListing = await createListing({
        statusID: ListingStatusID.AVAILABLE,
      });
      // all of these listings should not be listed
      await Promise.all(
        Object.values(ListingStatusID)
          .filter(
            (statusID) => typeof statusID === "number" && statusID !== ListingStatusID.AVAILABLE
          )
          .map((statusID) => createListing({ statusID }))
      );

      const listings = await service.list({
        agencyID: ModelFixtureConstants.verifiedAffiliateAgencyID,
      });
      expect(listings).toHaveLength(1);
      expect(listings[0].id).toBe(activeListing.id);
    });

    it("only lists listings that are available for affiliates", async () => {
      const listing1 = await createListing({
        visibilityID: ListingVisibilityID.AFFILIATE,
      });

      const listing2 = await createListing({
        visibilityID: ListingVisibilityID.STAFF_AFFILIATE,
      });

      // should not be listed
      await createListing({
        visibilityID: ListingVisibilityID.STAFF,
      });

      const listings = await service.list({
        agencyID: ModelFixtureConstants.verifiedAffiliateAgencyID,
      });
      expect(listings).toHaveLength(2);
      expect(listings[0].id).toBe(listing2.id);
      expect(listings[1].id).toBe(listing1.id);
    });

    it("only lists listings that are not allocated", async () => {
      const listing1 = await createListing({
        allocationID: ListingAllocationID.OPEN,
      });

      const listing2 = await createListing({
        allocationID: ListingAllocationID.INVESTOR_QUOTA,
      });

      // should not be listed
      await createListing({
        allocationID: ListingAllocationID.CLOSED,
      });

      const listings = await service.list({
        agencyID: ModelFixtureConstants.verifiedAffiliateAgencyID,
      });
      expect(listings).toHaveLength(2);
      expect(listings[0].id).toBe(listing2.id);
      expect(listings[1].id).toBe(listing1.id);
    });

    it("lists listings ordered by unit number", async () => {
      const listing1 = await createListing({
        unitNumber: "B1",
      });

      const listing2 = await createListing({
        unitNumber: "A2",
      });

      const listings = await service.list({
        agencyID: ModelFixtureConstants.verifiedAffiliateAgencyID,
      });

      expect(listings).toHaveLength(2);
      expect(listings[0].unitNumber).toBe(listing2.unitNumber);
      expect(listings[1].unitNumber).toBe(listing1.unitNumber);
    });

    it("translates purpose ID into a string", async () => {
      const listingForSale = await createListing({
        purposeID: PurposeID.FOR_SALE,
      });

      const listingForRent = await createListing({
        purposeID: PurposeID.FOR_RENT,
      });

      const listings = await service.list({
        agencyID: ModelFixtureConstants.verifiedAffiliateAgencyID,
      });

      expect(listings).toHaveLength(2);

      expect(listings[0].id).toBe(listingForRent.id);
      expect(listings[0].purpose).toBe(Purpose.FOR_RENT);
      expect(listings[1].id).toBe(listingForSale.id);
      expect(listings[1].purpose).toBe(Purpose.FOR_SALE);
    });

    each(["", null, undefined]).it("returns null for url if is %s", async (url) => {
      const listing = await createListing({
        url,
      });

      const listings = await service.list({
        agencyID: ModelFixtureConstants.verifiedAffiliateAgencyID,
      });

      expect(listings).toHaveLength(1);

      expect(listings[0].id).toBe(listing.id);
      expect(listings[0].url).toBe(null);
    });

    it("returns null for area if no data is available", async () => {
      const listing = await createListing({
        areaCovered: null,
        areaFree: null,
        areaTotal: null,
      });

      const listings = await service.list({
        agencyID: ModelFixtureConstants.verifiedAffiliateAgencyID,
      });
      expect(listings[0].id).toBe(listing.id);
      expect(listings[0].area).toBe(null);
    });

    it("computes total area if not available", async () => {
      const listing = await createListing({
        areaUnitID: AreaUnitID.KANAL,
        areaCovered: 10,
        areaFree: 5,
      });

      const listings = await service.list({
        agencyID: ModelFixtureConstants.verifiedAffiliateAgencyID,
      });
      expect(listings[0].id).toBe(listing.id);
      expect(listings[0].area).toStrictEqual({
        unit: AreaUnit.KANAL,
        value: {
          covered: 10,
          free: 5,
          total: 15,
        },
      });
    });

    it("assumes the area unit is sqft if not available", async () => {
      const listing = await createListing({
        areaUnitID: null,
        areaCovered: 10,
        areaFree: 5,
      });

      const listings = await service.list({
        agencyID: ModelFixtureConstants.verifiedAffiliateAgencyID,
      });
      expect(listings[0].id).toBe(listing.id);
      expect(listings[0].area?.unit).toBe(AreaUnit.SQFT);
    });

    it("returns null for price if not available", async () => {
      const listing = await createListing({
        price: null,
      });

      const listings = await service.list({
        agencyID: ModelFixtureConstants.verifiedAffiliateAgencyID,
      });
      expect(listings[0].id).toBe(listing.id);
      expect(listings[0].price).toBe(null);
    });

    it("does not convert 0 for bed and bathroom count into null", async () => {
      const listing = await createListing({
        bedroomCount: 0,
        bathroomCount: 0,
      });

      const listings = await service.list({
        agencyID: ModelFixtureConstants.verifiedAffiliateAgencyID,
      });
      expect(listings[0].id).toBe(listing.id);
      expect(listings[0].bedroomCount).toBe(0);
      expect(listings[0].bathroomCount).toBe(0);
    });

    it("assumes the commission structure is percentage if unknown", async () => {
      const listing = await createListing({
        commissionValue: 2.5,
      });

      const listings = await service.list({
        agencyID: ModelFixtureConstants.verifiedAffiliateAgencyID,
      });
      expect(listings[0].id).toBe(listing.id);
      expect(listings[0].commission).toStrictEqual({
        type: ListingCommissionType.PERCENTAGE,
        value: 2.5,
      });
    });

    it("returns only active images", async () => {
      const listing = await createListing();
      const mediaBankImage1 = await createMediaBankImage();
      const listingImage1 = await createListingImage(listing.id, mediaBankImage1.id, {
        isDeleted: false,
      });
      const mediaBankImage2 = await createMediaBankImage();
      await createListingImage(listing.id, mediaBankImage2.id, { isDeleted: true });

      const listings = await service.list({
        agencyID: ModelFixtureConstants.verifiedAffiliateAgencyID,
      });
      expect(listings[0].id).toBe(listing.id);
      expect(listings[0].images).toHaveLength(1);
      expect(listings[0].images[0].id).toBe(listingImage1.id);
      expect(listings[0].images).toMatchSnapshot();
    });

    it("returns only images with a valid media bank image", async () => {
      const listing = await createListing();
      await createListingImage(listing.id, null);

      const listings = await service.list({
        agencyID: ModelFixtureConstants.verifiedAffiliateAgencyID,
      });
      expect(listings[0].id).toBe(listing.id);
      expect(listings[0].images).toHaveLength(0);
    });

    each([
      [
        "locationID",
        {
          locationID: 6,
          countryLocationID: 1,
          regionLocationID: 2,
          cityLocationID: 3,
          localityLocationID: 4,
          phaseLocationID: 5,
        },
      ],
      [
        "phaseLocationID",
        {
          countryLocationID: 1,
          regionLocationID: 2,
          cityLocationID: 3,
          localityLocationID: 4,
          phaseLocationID: 5,
        },
      ],
      [
        "localityLocationID",
        { countryLocationID: 1, regionLocationID: 2, cityLocationID: 3, localityLocationID: 4 },
      ],
      ["cityLocationID", { countryLocationID: 1, regionLocationID: 2, cityLocationID: 3 }],
      ["regionLocationID", { countryLocationID: 1, regionLocationID: 2 }],
      ["countryLocationID", { countryLocationID: 1 }],
    ]).it("returns address's %s of the leaf location", async (addressFieldName, addressFields) => {
      const address = await Address.create(addressFields);
      const expectedLocationID = addressFields[addressFieldName];

      const listing = await createListing({ addressID: address.id });
      const listings = await service.list({
        agencyID: ModelFixtureConstants.verifiedAffiliateAgencyID,
      });
      expect(listings[0].id).toBe(listing.id);
      expect(listings[0].locationID).toBe(expectedLocationID);
    });

    it("returns null for locationID if no ID can be found in the address", async () => {
      const address = await Address.create({});

      const listing = await createListing({ addressID: address.id });
      const listings = await service.list({
        agencyID: ModelFixtureConstants.verifiedAffiliateAgencyID,
      });
      expect(listings[0].id).toBe(listing.id);
      expect(listings[0].locationID).toBe(null);
    });

    it("correctly filters by price if price filters are given", async () => {
      await createListing({ price: 5 });
      const listing = await createListing({ price: 10 });
      await createListing({ price: 15 });
      const listings = await service.list({
        agencyID: ModelFixtureConstants.verifiedAffiliateAgencyID,
        minPrice: 6,
        maxPrice: 11,
      });
      expect(listings).toBeTruthy();
      expect(listings).toHaveLength(1);
      expect(listings[0].id).toBe(listing.id);
    });
    it("correctly filters by area if area filters are given", async () => {
      await createListing({ areaTotal: 10 });
      const listing = await createListing({ areaTotal: 15 });
      await createListing({ areaTotal: 20 });
      const listings = await service.list({
        agencyID: ModelFixtureConstants.verifiedAffiliateAgencyID,
        minArea: 11,
        maxArea: 17,
      });
      expect(listings).toBeTruthy();
      expect(listings).toHaveLength(1);
      expect(listings[0].id).toBe(listing.id);
    });
  });
  it("correctly filters by categoryIDs if categoryIDs are given", async () => {
    await Promise.all([1, 1, 2, 3, 4].map((categoryID: number) => createListing({ categoryID })));
    const listings = await service.list({
      agencyID: ModelFixtureConstants.verifiedAffiliateAgencyID,
      categoryIDs: [1, 2],
    });
    expect(listings).toBeTruthy();
    expect(listings).toHaveLength(3);
    // finds both listings with same categoryID 1
    expect(listings.filter((listing: ListingDTO) => listing.categoryID === 1)).toHaveLength(2);
    // no listing is found outside of the filter criterion
    expect(
      listings.filter((listing: ListingDTO) => listing.categoryID === 3 || listing.categoryID === 4)
    ).toHaveLength(0);
  });

  describe("statsGroupedBy", () => {
    it("returns stats grouped by", async () => {
      const againstID1 = 1337;
      const againstID2 = 7331;

      await createListing({ againstID: againstID1, categoryID: 4, price: 400 });
      await createListing({ againstID: againstID1, categoryID: 3, price: 300 });
      await createListing({ againstID: againstID2, categoryID: 3, price: 500 });

      const stats = await service.statsGroupedBy({
        agencyID: ModelFixtureConstants.verifiedAffiliateAgencyID,
      });
      expect(stats.size).toBe(2);
      expect(stats.get(againstID1)).toStrictEqual({
        count: 2,
        categoryIDs: [4, 3],
        minPrice: 300,
        maxPrice: 400,
      });
      expect(stats.get(againstID2)).toStrictEqual({
        count: 1,
        categoryIDs: [3],
        minPrice: 500,
        maxPrice: 500,
      });
    });

    it("only uses open listings", async () => {
      const againstID = 1;

      await createListing({ againstID, categoryID: 3, price: 400 });
      await createListing({
        againstID,
        statusID: ListingStatusID.CLOSED,
        price: 500,
        categoryID: 4,
      });

      const stats = await service.statsGroupedBy({
        agencyID: ModelFixtureConstants.verifiedAffiliateAgencyID,
      });
      expect(stats.size).toBe(1);
      expect(stats.get(againstID)).toStrictEqual({
        count: 1,
        categoryIDs: [3],
        minPrice: 400,
        maxPrice: 400,
      });
    });

    it("only uses listings matching the agency ID", async () => {
      const againstID = 1;

      await createListing({
        agencyID: ModelFixtureConstants.verifiedAffiliateAgencyID,
        againstID,
        categoryID: 3,
        price: 400,
      });
      await createListing({
        agencyID: 999,
        againstID,
        price: 500,
        categoryID: 4,
      });

      const stats = await service.statsGroupedBy({
        agencyID: ModelFixtureConstants.verifiedAffiliateAgencyID,
      });
      expect(stats.size).toBe(1);
      expect(stats.get(againstID)).toStrictEqual({
        count: 1,
        categoryIDs: [3],
        minPrice: 400,
        maxPrice: 400,
      });
    });

    it("only uses listings available for affiliates", async () => {
      const againstID = 1;

      await createListing({ againstID, categoryID: 3, price: 400 });
      await createListing({
        againstID,
        visibilityID: ListingVisibilityID.STAFF,
        price: 500,
        categoryID: 4,
      });

      const stats = await service.statsGroupedBy({
        agencyID: ModelFixtureConstants.verifiedAffiliateAgencyID,
      });
      expect(stats.size).toBe(1);
      expect(stats.get(againstID)).toStrictEqual({
        count: 1,
        categoryIDs: [3],
        minPrice: 400,
        maxPrice: 400,
      });
    });

    it("only uses non-closed listings", async () => {
      const againstID = 1;

      await createListing({ againstID, categoryID: 3, price: 400 });
      await createListing({
        againstID,
        allocationID: ListingAllocationID.CLOSED,
        price: 500,
        categoryID: 4,
      });

      const stats = await service.statsGroupedBy({
        agencyID: ModelFixtureConstants.verifiedAffiliateAgencyID,
      });
      expect(stats.size).toBe(1);
      expect(stats.get(againstID)).toStrictEqual({
        count: 1,
        categoryIDs: [3],
        minPrice: 400,
        maxPrice: 400,
      });
    });
  });

  describe("getSoldListingsCount", () => {
    it("returns soldCount grouped by againstID", async () => {
      const againstID1 = 1;
      const againstID2 = 7331;

      await loadSequelizeFixtures(
        ModelFixtureNames.LISTINGS,
        ModelFixtureNames.CONTRACT_STATUSES,
        ModelFixtureNames.CONTRACTS,
        ModelFixtureNames.CONTRACT_LISTINGS
      );

      const soldCount = await service.getSoldListingsCount(
        ModelFixtureConstants.verifiedAffiliateAgencyID,
        [againstID1, againstID2]
      );
      expect(soldCount.size).toBe(1);
      expect(soldCount.get(againstID1)).toStrictEqual(2);
      expect(soldCount.get(againstID2)).toBeUndefined();
    });

    it("only uses listings matching the agency ID", async () => {
      const againstID = 1;

      await loadSequelizeFixtures(
        ModelFixtureNames.LISTINGS,
        ModelFixtureNames.CONTRACT_STATUSES,
        ModelFixtureNames.CONTRACTS,
        ModelFixtureNames.CONTRACT_LISTINGS
      );

      const soldCount = await service.getSoldListingsCount(
        ModelFixtureConstants.verifiedAffiliateAgencyID,
        [againstID]
      );
      expect(soldCount.size).toBe(1);
      expect(soldCount.get(againstID)).toStrictEqual(2);
    });

    it("it only uses those listings having againstID that is provided", async () => {
      const againstID1 = 1;
      const againstID2 = 5;

      await loadSequelizeFixtures(
        ModelFixtureNames.LISTINGS,
        ModelFixtureNames.CONTRACT_STATUSES,
        ModelFixtureNames.CONTRACTS,
        ModelFixtureNames.CONTRACT_LISTINGS
      );

      const soldCount = await service.getSoldListingsCount(
        ModelFixtureConstants.verifiedAffiliateAgencyID,
        [againstID1]
      );
      expect(soldCount.size).toBe(1);
      expect(soldCount.get(againstID1)).toStrictEqual(2);
      expect(soldCount.get(againstID2)).toBeUndefined();
    });
  });

  describe("find", () => {
    it("successfully finds active listing visible to affiliates", async () => {
      const listing = await createListing({
        statusID: ListingStatusID.AVAILABLE,
        allocationID: ListingAllocationID.OPEN,
        visibilityID: ListingVisibilityID.STAFF_AFFILIATE,
        agencyID: ModelFixtureConstants.verifiedAffiliateAgencyID,
      });
      const foundListing = await service.find({
        listingID: listing.id,
        agencyID: ModelFixtureConstants.verifiedAffiliateAgencyID,
      });

      expect(foundListing).toBeTruthy();
      expect(foundListing?.id).toBe(listing.id);
    });

    it("does not find listing with invalid id", async () => {
      const listing = await createListing();

      const foundListing = await service.find({
        listingID: listing.id + 999,
        agencyID: ModelFixtureConstants.verifiedAffiliateAgencyID,
      });

      expect(foundListing).toBeNull();
    });

    it("does not find reserved listing", async () => {
      const listing = await createListing({ statusID: ListingStatusID.RESERVED_HOLD });

      const foundListing = await service.find({
        listingID: listing.id,
        agencyID: ModelFixtureConstants.verifiedAffiliateAgencyID,
      });

      expect(foundListing).toBeNull();
    });
  });
});
