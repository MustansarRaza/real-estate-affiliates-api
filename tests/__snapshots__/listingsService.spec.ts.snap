// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListingsService list returns only active images 1`] = `
Array [
  Object {
    "category": "photo",
    "displayOrder": 0,
    "id": 1,
    "url": https://s3bucket.s3.eu-west-1.amazonaws.com/image.jpg?AWSAccessKeyId=S3ACCESSKEYID&Expires=EXPIRES&Signature=SIGNATURE,
  },
]
`;
